test/e2e/network/service.go

```diff
@@ -2730,6 +2730,98 @@ var _ = common.SIGDescribe("Services", func() {
 		framework.ExpectNoError(err, "failed to delete Service %v in namespace %v", testService.ObjectMeta.Name, ns)
 		framework.Logf("Service %s deleted", testSvcName)
 	})
+
+	ginkgo.It("should delete a collection of services", func() {
+
+		ns := f.Namespace.Name
+		svcClient := f.ClientSet.CoreV1().Services(ns)
+		svcResource := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "services"}
+		svcDynamicClient := f.DynamicClient.Resource(svcResource).Namespace(ns)
+
+		svcName := "e2e-svc-" + utilrand.String(5)
+		svcLabels := map[string]string{"e2e-test-service": "true"}
+		labelSelector := labels.SelectorFromSet(svcLabels).String()
+
+		w := &cache.ListWatch{
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.LabelSelector = labelSelector
+				return svcClient.Watch(context.TODO(), options)
+			},
+		}
+
+		svcList, err := cs.CoreV1().Services("").List(context.TODO(), metav1.ListOptions{LabelSelector: labelSelector})
+		framework.ExpectNoError(err, "failed to list Services")
+
+		ginkgo.By("creating a Service")
+		testService := v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   svcName,
+				Labels: svcLabels,
+			},
+			Spec: v1.ServiceSpec{
+				Type: "ClusterIP",
+				Ports: []v1.ServicePort{{
+					Name:       "http",
+					Protocol:   v1.ProtocolTCP,
+					Port:       int32(80),
+					TargetPort: intstr.FromInt(80),
+				}},
+			},
+		}
+		_, err = svcClient.Create(context.TODO(), &testService, metav1.CreateOptions{})
+		framework.ExpectNoError(err, "failed to create Service")
+
+		ginkgo.By("watching for the Service to be added")
+		ctx, cancel := context.WithTimeout(context.Background(), svcReadyTimeout)
+		defer cancel()
+		_, err = watchtools.Until(ctx, svcList.ResourceVersion, w, func(event watch.Event) (bool, error) {
+			if svc, ok := event.Object.(*v1.Service); ok {
+				found := svc.ObjectMeta.Name == testService.ObjectMeta.Name &&
+					svc.ObjectMeta.Namespace == ns &&
+					svc.Labels["e2e-test-service"] == "true"
+				if !found {
+					framework.Logf("observed Service %v in namespace %v with labels: %v & ports %v", svc.ObjectMeta.Name, svc.ObjectMeta.Namespace, svc.Labels, svc.Spec.Ports)
+					return false, nil
+				}
+				framework.Logf("Found Service %v in namespace %v with labels: %v & ports %v", svc.ObjectMeta.Name, svc.ObjectMeta.Namespace, svc.Labels, svc.Spec.Ports)
+				return found, nil
+			}
+			framework.Logf("Observed event: %+v", event.Object)
+			return false, nil
```

Comment:
can this happen? should be a reason to fail the test?

Reply:
Because we are watching events before the service was created we will see when the service was created (ADDED)

```
 [It] should delete a collection of services
   /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/service.go:2734
 STEP: creating a Service
 STEP: watching for the Service to be added
 Nov  1 03:34:25.572: INFO: Found Service e2e-svc-hbqm8 in namespace services-5872 with labels: map[e2e-test-service:true] & ports [{http TCP <nil> 80 {0 80 } 0}]
 Nov  1 03:34:25.572: INFO: Service e2e-svc-hbqm8 created
 STEP: deleting service collection
 STEP: watching for the Service to be deleted
 Nov  1 03:34:25.603: INFO: Observed event: ADDED
 Nov  1 03:34:25.604: INFO: Found Service e2e-svc-hbqm8 in namespace services-5872 with labels: map[e2e-test-service:true] & annotations: map[]
 Nov  1 03:34:25.604: INFO: Service e2e-svc-hbqm8 deleted
``